### DISCUSSION TITLE
추천 시스템 로직

### DISCUSSION THREAD

**Discussion by kang20** [LikeLion12th-OMEB/Backend#33] [Q&A] [answered]
![image](https://github.com/user-attachments/assets/77a505da-df9e-4f31-b6e5-f175700de916)
### event_review

```java
{
	"user_id":"",
	"book_id":"",
	"emotion":"SADNESS",
	"time":"",
}
```

### event_view

```java
{
	"user_id":"",
	"book_id":"",
	"time":"",
}
```

### event_recommanded

```java
{
	"user_id":"",
	"book_id":"",
	"emotion":"SADNESS",
	"time":"",
}
```

### user_score

```java
{
  "_id": "book1",
  "user_id":""
  "sadness": 0.5,
  "anger": 1.0,
  "anxiety": 0.8,
  "loneliness": 0.7,
  "jealousy": 0.4,
  "happiness": 3.0,
  "lethargy": 0.2,
  "love": 4.5,
  "sense_of_achievement": 2.3
}
```

### book_score

```java
{
  "_id": "book1",
	"book_id":"",
  "sadness": 0.5,
  "anger": 1.0,
  "anxiety": 0.8,
  "loneliness": 0.7,
  "jealousy": 0.4,
  "happiness": 3.0,
  "lethargy": 0.2,
  "love": 4.5,
  "sense_of_achievement": 2.3
}
```

### emotion_rank

```java
{
  "_id": "emotion_id",  // 예: "ANGRY"
  "emotion": "ANGRY",
  "books": [
    {
      "book_id": "book1",
      "title": "Book Title 1",
      "image_url": "http://example.com/image1.jpg",
      "price": 20.99,
      "author": "Author Name",
      "rank": 1,
      "score": 95.5
    },
    {
      "book_id": "book2",
      "title": "Book Title 2",
      "image_url": "http://example.com/image2.jpg",
      "price": 15.99,
      "author": "Author Name",
      "rank": 2,
      "score": 90.0
    },
    ...
  ]
}
6
```

## Batch 작업

- 정해진 시간에 작업
- redis에 쌓여있는 event 로그를 기반으로 user_score 갱신
- book_score 갱신
- book 갱신 기준으로 emotion_rank 변동 확인
- 갱신할때 최신 score의 점유를 크게하기 위해 감마를 곱해서 감쇠할 예정
- user_score 갱신은 review , view ,recommand 기준으로 갱신
    - view는 emotion 필드가 없는대신 book 의 주 감정(제일 점수 높은 감정)으로 갱신할 예정
- book_score 갱신은 review , view 기준으로 갱신
    - view 는 emotion 필드가 없는 대신 view 한 user의 주 감정으로 갱신할 예정

**Comment by siwan9** [comment, marked as answer]
user_score의 이유가 책 조회에 따른 book_score 갱신을 위해 존재하는 것으로 보이는데, user_score로 집계된 사용자의 주 감정이 책 조회에 반영되는 감정과 일치될 가능성은 높지만 애매하므로 user_score를 삭제하고 단순하게 책 조회 시 책의 특정 감정의 점수를 올려주는 방법은 어떠한가요?
